### PR TITLE
WIP: Aarch64 eltwise Binary and Unary simple ops enabling

### PR DESCRIPTION
**Pull request by deeptiag1**
Fixed CI issues.
Enabled Binary and Unary simple ops

### CODE REVIEW

File: src/generator_mateltwise_aarch64.c [hunk 1]
```diff
@@ -33,18 +33,10 @@ void libxsmm_generator_mateltwise_aarch64_update_micro_kernel_config_vectorlengt
       io_micro_kernel_config->datatype_size_in = 4;
       io_micro_kernel_config->vector_length_in = 4;
       io_micro_kernel_config->vmove_instruction_in = LIBXSMM_AARCH64_INSTR_ASIMD_LDR_R;
-    } else if ( LIBXSMM_DATATYPE_BF16 == LIBXSMM_GETENUM_INP( i_mateltwise_desc->datatype ) ) {
```
Comment (alheinecke) [inline src/generator_mateltwise_aarch64.c:36]: Please add this back

File: src/generator_mateltwise_aarch64.c [hunk 2]
```diff
@@ -33,18 +33,10 @@ void libxsmm_generator_mateltwise_aarch64_update_micro_kernel_config_vectorlengt
...
-      io_micro_kernel_config->datatype_size_in = 2;
-      io_micro_kernel_config->vector_length_in = 8;
-      io_micro_kernel_config->vmove_instruction_in = LIBXSMM_AARCH64_INSTR_ASIMD_LDR_R;
     } else if ( LIBXSMM_DATATYPE_F16 == LIBXSMM_GETENUM_INP( i_mateltwise_desc->datatype ) ) {
       io_micro_kernel_config->datatype_size_in = 2;
       io_micro_kernel_config->vector_length_in = 8;
       io_micro_kernel_config->vmove_instruction_in = LIBXSMM_AARCH64_INSTR_ASIMD_LDR_R;
-    } else if ( LIBXSMM_DATATYPE_I16 == LIBXSMM_GETENUM_INP( i_mateltwise_desc->datatype ) ) {
```
Comment (alheinecke) [inline src/generator_mateltwise_aarch64.c:44]: Please add this back

File: src/generator_mateltwise_aarch64.c [hunk 3]
```diff
@@ -62,18 +54,10 @@ void libxsmm_generator_mateltwise_aarch64_update_micro_kernel_config_vectorlengt
       io_micro_kernel_config->datatype_size_out = 4;
       io_micro_kernel_config->vector_length_out = 4;
       io_micro_kernel_config->vmove_instruction_out = LIBXSMM_AARCH64_INSTR_ASIMD_LDR_R;
-    } else if ( LIBXSMM_DATATYPE_BF16 == LIBXSMM_GETENUM_OUT( i_mateltwise_desc->datatype ) ) {
```
Comment (alheinecke) [inline src/generator_mateltwise_aarch64.c:65]: Please add this back

File: src/generator_mateltwise_aarch64.c [hunk 4]
```diff
@@ -62,18 +54,10 @@ void libxsmm_generator_mateltwise_aarch64_update_micro_kernel_config_vectorlengt
...
-      io_micro_kernel_config->datatype_size_out = 2;
-      io_micro_kernel_config->vector_length_out = 8;
-      io_micro_kernel_config->vmove_instruction_out = LIBXSMM_AARCH64_INSTR_ASIMD_LDR_R;
     } else if ( LIBXSMM_DATATYPE_F16 == LIBXSMM_GETENUM_OUT( i_mateltwise_desc->datatype ) ) {
       io_micro_kernel_config->datatype_size_out = 2;
       io_micro_kernel_config->vector_length_out = 8;
       io_micro_kernel_config->vmove_instruction_out = LIBXSMM_AARCH64_INSTR_ASIMD_LDR_R;
-    } else if ( LIBXSMM_DATATYPE_I16 == LIBXSMM_GETENUM_OUT( i_mateltwise_desc->datatype ) ) {
```
Comment (alheinecke) [inline src/generator_mateltwise_aarch64.c:73]: Please add this back

File: src/generator_mateltwise_aarch64.c [hunk 5]
```diff
@@ -221,24 +211,7 @@ void libxsmm_generator_mateltwise_aarch64_kernel( libxsmm_generated_code*
     /* Stack management for melt kernel */
     libxsmm_generator_meltw_setup_stack_frame_aarch64( io_generated_code, i_mateltwise_desc, &l_gp_reg_mapping, &l_kernel_config);
 
-    if (i_mateltwise_desc->operation == LIBXSMM_MELTW_OPERATION_UNARY ) {
```
Comment (alheinecke) [inline src/generator_mateltwise_aarch64.c:224]: Please revert this change, you also need to add a branch for Binary... please have a look at the x86 code as template

Comment (alheinecke) [inline src/generator_mateltwise_aarch64.c:224]: Please revert this change, you also need to add a branch for Binary... please have a look at the x86 code as template

File: src/generator_mateltwise_unary_binary_aarch64.c [hunk 6]
```diff
@@ -363,19 +645,34 @@ void libxsmm_load_aarch64_2d_reg_block( libxsmm_generated_code*
...
-    }
+      if ( bcast_input == 0 ) {
+        if ( l_ld_bytes-(16*i_m_blocking) > 0 ) {
+          libxsmm_aarch64_instruction_alu_compute_imm64( io_generated_code, LIBXSMM_AARCH64_INSTR_GP_META_ADD,
+                                                        i_gp_reg_mapping->gp_reg_in, i_gp_reg_mapping->gp_reg_scratch_0, i_gp_reg_mapping->gp_reg_in,
+                                                        (unsigned long long)((unsigned long long)(l_ld_bytes-(16*i_m_blocking))) );
+        }
+      }
+  }
+  if( bcast_input == 0 ) {
```
Comment (alheinecke) [inline src/generator_mateltwise_unary_binary_aarch64.c:661]: please wrap the offset calculation

File: src/generator_mateltwise_unary_binary_aarch64.c [hunk 7]
```diff
@@ -1236,7 +1556,192 @@ void libxsmm_compute_unary_binary_aarch64_2d_reg_block( libxsmm_generated_code*
...
+
+
+    }
+    if ( l_ld_bytes-(16*i_m_blocking) > 0 ) {
+      libxsmm_aarch64_instruction_alu_compute_imm64( io_generated_code, LIBXSMM_AARCH64_INSTR_GP_META_ADD,
+                                                     i_gp_reg_mapping->gp_reg_in2, i_gp_reg_mapping->gp_reg_scratch_0, i_gp_reg_mapping->gp_reg_in2,
+                                                     (unsigned long long)((unsigned long long)(l_ld_bytes-(16*i_m_blocking))) );
+    }
+  }
+  if( ( bcast_input == 0) || (( bcast_col == 1 ) & (in == 0)) ) {
```
Comment (alheinecke) [inline src/generator_mateltwise_unary_binary_aarch64.c:1721]: Please wrap the offset calculation

**Review by alheinecke — CHANGES_REQUESTED** [review]
It seems in generator_mateltwise_aarch64.c merge conflicts have been handled incorrectly that led to disabling the transposes and binary was not added properly.

It seems this was not found as the unit tests have not been extended to all the binary & unary functions which are added with this PR. Please add the unit tests (ala copy and transpose in kernel_test_aarch64, can be looked at how copy was created from the x86 test) and add all of these tests to run_test_aarch64.sh

**Comment by alheinecke** [comment]
code looks good, please add the unit tests, check them and then we can merge

File: src/generator_mateltwise_unary_binary_aarch64.c (hunk 6 above)
Comment (alheinecke) [reply src/generator_mateltwise_unary_binary_aarch64.c:660]: this is still open I guess

**Review by alheinecke — COMMENTED** [review]
Are now all unit tests passing?

**Comment by deeptiag1** [comment]
FMULADD(Binary) and Unary(Inc) is left. Test pass for rest of them.

**Comment by alheinecke** [comment]
OK, let me know when done, then I can merge

**Review by alheinecke — APPROVED** [review]

**Pull request merged**